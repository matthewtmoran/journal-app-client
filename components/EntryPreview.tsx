import React from "react";
import { StyleSheet, View, TouchableOpacity } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { FlatList } from "react-native";
import IEntry from "../interfaces/IEntry";
import {
  RobotLightItalicText,
  RobotThinItalicText,
  RobotText,
} from "./StyledText";
import { format } from "date-fns";
import commonStyles from "../style/common";

interface IProps {
  entry: IEntry;
  index: number;
}

const EntryPreview = ({ entry, index }: IProps) => {
  const navigation = useNavigation();
  const handlePress = () => {
    navigation.navigate("EntryDetails", { entry });
  };

  const updatedAt = format(new Date(entry.updatedAt), "MMM do, yyyy");
  return (
    <TouchableOpacity
      style={styles.container}
      onPress={handlePress}
      testID="preview-button"
    >
      <View>
        <RobotText numberOfLines={1} style={styles.title}>
          {entry.title}
        </RobotText>
        <RobotLightItalicText style={styles.description}>
          {updatedAt}
        </RobotLightItalicText>
        <RobotThinItalicText numberOfLines={1} style={styles.description}>
          {entry.description}
        </RobotThinItalicText>
      </View>
      <FlatList
        style={styles.categories}
        data={entry.categories}
        keyExtractor={(item) => item.id!}
        listKey={`CategoryColorList-${index}`}
        renderItem={({ item }) => {
          return (
            <View
              testID={"category"}
              key={item.id}
              style={{
                ...commonStyles.categoryCircle,
                backgroundColor: `${item.color}`,
              }}
            />
          );
        }}
      />
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 100,
    marginVertical: 10,
    marginHorizontal: 10,
    flex: 1,
    padding: 8,
    borderRadius: 4,
    backgroundColor: "#fff",
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 3,
    display: "flex",
    justifyContent: "space-between",
  },
  categories: {
    display: "flex",
    flexDirection: "row",
    overflow: "scroll",
  },
  item: {
    padding: 10,
    fontSize: 18,
    height: 44,
  },
  title: {
    height: 18,
    fontSize: 16,
  },
  description: {
    fontSize: 14,
    overflow: "hidden",
  },
});

export default EntryPreview;
