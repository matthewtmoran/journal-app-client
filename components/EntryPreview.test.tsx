import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import EntryPreview from "./EntryPreview";

const mockNavigate = jest.fn();
jest.mock("@react-navigation/native", () => ({
  useNavigation: () => {
    return {
      navigate: mockNavigate,
    };
  },
}));

describe("components/EntryPreview", () => {
  const props = {
    entry: {
      id: "1234",
      title: "My-Title",
      updatedAt: "2020-06-17T21:25:04.436Z",
      createdAt: "2020-06-17T21:25:04.436Z",
      description: "This is a description",
      body: "This is the body",
      categories: [
        { id: "1", name: "thoughts", color: "#c1c1c1" },
        { id: "2", name: "feelings", color: "#333333" },
      ],
      imagePath: "",
      audioPath: "some/local/path/to/audio/file",
    },
    index: 1,
  };

  it("Should match snapshot", () => {
    const { baseElement } = render(<EntryPreview {...props}></EntryPreview>);
    expect(baseElement).toMatchSnapshot();
  });

  it("Should render title", async () => {
    const { findByText } = render(<EntryPreview {...props}></EntryPreview>);
    const titleText = await findByText("My-Title");
    expect(titleText).toBeTruthy();
  });

  it("Should render formatted date", async () => {
    const { findByText } = render(<EntryPreview {...props}></EntryPreview>);
    const text = await findByText("Jun 17th, 2020");
    expect(text).toBeTruthy();
  });

  it("Should render description", async () => {
    const { findByText } = render(<EntryPreview {...props}></EntryPreview>);
    const text = await findByText("This is a description");
    expect(text).toBeTruthy();
  });

  it("Should render categories", async () => {
    const { findAllByTestId } = render(
      <EntryPreview {...props}></EntryPreview>
    );
    const categories = await findAllByTestId("category");
    expect(categories).toHaveLength(2);
  });

  // it("Should call navigate with proper route", async () => {
  //   const { baseElement, getByTestId } = render(
  //     <EntryPreview {...props}></EntryPreview>
  //   );
  //   await fireEvent.press(getByTestId("preview-button"));
  //   expect(mockNavigate).toHaveBeenCalledWith("EntryDetails", {
  //     entry: props.entry,
  //   });
  // });
});
