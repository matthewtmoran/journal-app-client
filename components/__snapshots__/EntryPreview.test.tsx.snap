// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/EntryPreview Should match snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 4,
          "display": "flex",
          "elevation": 3,
          "flex": 1,
          "height": 100,
          "justifyContent": "space-between",
          "marginHorizontal": 10,
          "marginVertical": 10,
          "padding": 8,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.22,
          "shadowRadius": 2.22,
        }
      }
      testID="preview-button"
    >
      <View>
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontSize": 16,
                "height": 18,
              },
              Object {
                "fontFamily": "roboto",
              },
            ]
          }
        >
          My-Title
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "overflow": "hidden",
              },
              Object {
                "fontFamily": "roboto-light-italic",
              },
            ]
          }
        >
          Jun 17th, 2020
        </Text>
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontSize": 14,
                "overflow": "hidden",
              },
              Object {
                "fontFamily": "roboto-thin-italic",
              },
            ]
          }
        >
          This is a description
        </Text>
      </View>
      <ScrollView
        data={
          Array [
            Object {
              "color": "#c1c1c1",
              "id": "1",
              "name": "thoughts",
            },
            Object {
              "color": "#333333",
              "id": "2",
              "name": "feelings",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        listKey="CategoryColorList-1"
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        removeClippedSubviews={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "overflow": "scroll",
          }
        }
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "#c1c1c1",
                "borderRadius": 100,
                "height": 16,
                "margin": 2,
                "width": 16,
              }
            }
            testID="category"
          />
        </View>
        <View
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "#333333",
                "borderRadius": 100,
                "height": 16,
                "margin": 2,
                "width": 16,
              }
            }
            testID="category"
          />
        </View>
      </ScrollView>
    </TouchableOpacity>
  </View>
</View>
`;
